spatialdata/_core/_spatialdata.py

```diff
@@ -430,40 +448,52 @@ def add_image(
             See https://zarr.readthedocs.io/en/stable/api/storage.html for more details.
         overwrite
             If True, overwrite the element if it already exists.
-        _add_in_memory
+        _called_from_write
             Internal flag, to differentiate between an element added by the user and an element saved to disk by
             write method.
 
         Notes
         -----
         If the SpatialData object is backed by a Zarr storage, the image will be written to the Zarr storage.
         """
-        if _add_in_memory:
-            self._add_image_in_memory(name=name, image=image, overwrite=overwrite)
         if self.is_backed():
-            elem_group = self._init_add_element(name=name, element_type="images", overwrite=overwrite)
-            write_image(
-                image=self.images[name],
-                group=elem_group,
-                name=name,
-                storage_options=storage_options,
-            )
-            # reload the image from the Zarr storage so that now the element is lazy loaded, and most importantly,
-            # from the correct storage
+            self._add_image_in_memory(name=name, image=image, overwrite=overwrite)
+            with tempfile.TemporaryDirectory() as tmpdir:
```

Comment:
Why does adding an image cause any write to happen? Couldn't the image stay where it is, and we just hold a delayed representation here?

Then, when `write` is called explicitly, its written?

Reply:
that also a bit concerning to me, I'd rather have a warning saying that it's not backed/saved rather than creating a temporary dir

Reply:
@ivirshup the motivation to writing to disk is the following: if a spatialdata object is backed (=it has been written to disk or it has read from disk; note, when I write I immediately re-read to address [this problem](https://github.com/scverse/spatialdata-io/issues/12)), then every time I add something to it: an image, the table, a transformation, I immediately save it to disk; I believe that this makes the life easier to the user, and for us, because we don't have to keep track to what has been saved and what not

@giovp calling write explicitly would not remove the need for the temporary directory. The temporary directory is for performing an atomic save. In this way if the destination of the save is the same as the origin of the backed data, we can still perform the save operation.

Reply:
Consider the following example:

```
sdata = read_zarr(...)
sdata.images['my_image'] = dask.array.log1p(sdata.images['my_image']
```

sdata.images['my_image'] is never loaded in memory, but it saved chunk by chunk from dask. This operation is possible because of the atomic save, otherwise we would not be able to modify the group. This issue is what Wouter reported.

Reply:
So `add_image` causes the array represented by `normalized` to compute and write to disk?

I think that's not great, and skips out on a lot of what makes dask useful.

I assume that would run when either `sdata.write(...)` or `sdata.compute(...)` was called.

Reply:
> atomic save

Atomicity would be great, but also is a high bar and not something I think is in scope for 0.1. TBH, I would say "we should never aim for atomic updates with `zarr.DirectoryStore`". I think we could maybe (at some point in the future) do that with a kerchunk or iceberg style store.